### PR TITLE
feat(agents)!: per-agent model selection for cost optimization and /compact loop fix

### LINKED ISSUE #1420: Autopilot mode in VS Code auto‑executes first handoff (/compact) → infinite loop in hve-core custom agents
### Component

Other

### Autopilot Behavior

When **Autopilot mode** is enabled in VS Code GitHub Copilot:
- Copilot automatically selects and executes the **first available handoff**
- It does **not wait for user interaction or selection**

---

### Loop Scenario with `/compact`

If the **first handoff is `/compact`**:
- The agent re-enters itself via the handoff
- Completes execution and triggers the same handoff again
- This cycle repeats indefinitely

➡️ Result: **Infinite loop of `/compact` execution**

Image

## Expected Behavior

After completing its execution, a custom agent should:

- Finish the current workflow
- Present available **handoffs as suggestions**
- Wait for **explicit user interaction** before executing any handoff

This behavior works correctly in:
- **Default mode**
- **Bypass Approvals mode**

In these modes, handoffs are displayed as options and require user selection before execution.

### Autopilot Mode (Expected)

In **Autopilot mode**, the expected behavior should be consistent:

- The agent should complete execution
- Handoffs should be surfaced as **suggested next actions**
- Copilot should **not automatically execute any handoff**
- Execution should pause or await user input unless explicitly configured otherwise

### Observed Deviation

In Autopilot mode:

- Copilot automatically selects and executes the **first available handoff**
- No user interaction is required
- If the first handoff is `/compact` and points back to the same agent:
  - The agent re-enters execution
  - Completes again
  - Triggers the same handoff
  - Repeats indefinitely

This leads to an **infinite loop**, breaking the expected user-driven handoff model.
``

## Steps to Reproduce

1. Enable **"Autopilot" mode** in VS Code Copilot Chat.
2. Use any **custom agent** with handoffs configured (e.g., `RPI Agent` or similar from `hve-core`).
3. Let the agent run to completion.
4. Observe behavior after completion:
   - Copilot **automatically executes the first handoff** instead of waiting for user input.
   - If the first handoff is `/compact`, the agent re-enters execution.
   - This repeats indefinitely, resulting in an **infinite loop**.

## Additional Context

- I’m not sure if this is the intended behavior in **Autopilot** mode (since it’s in preview), and whether there are plans to address it.

### Initial Guess / RCA (Hypothesis)

- Autopilot runs the custom agent normally.
- Once the agent finishes, it inspects the configured **handoff steps**.
  - Ideally, these handoffs should be presented as **recommendations/options** for the user to choose from.
  - However, in Autopilot, it appears to **automatically select and execute the first handoff**.
- If the first handoff is something like **Compact** (and it routes back into the same agent / same completion path), this creates a self-reinforcing cycle:
  - agent completes → handoffs evaluated → first handoff auto-executed → agent re-enters → completes again → repeat  
  → resulting in an **infinite loop**.

### Validation / What We Observed

- The exact same agents work correctly in:
  - **Default** mode
  - **Bypass Approvals** mode  
  In these modes, the agent completes and **handoffs are shown as options**, waiting for user selection.

- In **Autopilot** mode:
  - If an agent **has handoffs configured**, it can end up in a loop (especially when the first handoff is `/compact` or routes back into the same agent flow).
  - If an agent **does NOT have any handoffs configured** (example: `arch-diagram-builder`), it does **not** loop.
  
This strongly suggests the issue is specifically tied to how **Autopilot handles handoffs** (auto-execution of first option) rather than the agent logic itself.

### LINKED ISSUE #1540: feat(agents): per-agent model selection for cost optimization
## Summary

Introduce per-agent and per-prompt model selection via frontmatter, backed by a validated model catalog that stays current as GitHub Copilot's model lineup evolves (~every 2 months).

## Motivation

Both VS Code and the GitHub CLI now support per-request model selection. This unlocks cost and performance optimization: simple prompts (commit messages, git setup, issue triage) don't need expensive reasoning models - a fast-tier model handles them in fewer tokens at a fraction of the cost. Complex agents (planning, implementation, security review) benefit from premium models that justify their multiplier. Without explicit selection, every request defaults to the same model regardless of task complexity, wasting budget on trivial tasks and potentially under-serving complex ones.

Additionally, Copilot's supported model set changes frequently (~every 2 months): new models arrive, existing models get renamed or retired, and pricing tiers shift. Without validation infrastructure, stale model references silently degrade agent behavior or cause fallback to defaults. Agents and prompts should declare preferred models explicitly, and the tooling should catch drift before it reaches users.

## Scope

* `model` frontmatter property on `.agent.md` and `.prompt.md` files (string or ordered array for fallback)
* Reference catalog (`model-catalog.json`) with name, tier, multiplier, and status per model
* Validation script that flags invalid or retiring model references at lint time
* Automated catalog refresh from GitHub's authoritative YAML data sources
* Weekly CI check to detect model additions, removals, and renames upstream

## Key Design Decisions

* Catalog is dynamically refreshed from `github/docs` structured YAML (not scraped from rendered pages)
* Removed models are marked `retiring` with a 60-day grace period rather than deleted immediately
* Validation produces warnings for retiring references and errors for unknown references
* The refresh runs in CI weekly and can be triggered manually; no runtime network calls during validation

## Maintenance Expectations

GitHub's model lineup typically changes every 6-8 weeks. The automated refresh handles additions and multiplier updates without manual intervention. Human review is needed when:

* A model referenced in agent files is retired (validation warns, PR review catches)
* The upstream YAML schema changes (refresh script would fail, CI alerts)
* New tier boundaries need adjustment (rare, requires script update)

## Acceptance Criteria

* [ ] Agent and prompt files declare model preferences via frontmatter
* [ ] `npm run lint:models` validates all references against the catalog
* [ ] `npm run lint:models:refresh` updates the catalog from upstream sources
* [ ] Weekly CI workflow detects catalog drift and reports retiring models
* [ ] Pester tests cover validation logic and catalog merge logic at 95%+ coverage
* [ ] Retiring model warnings surface in PR checks before references break
```
## Notes

* 14 files have model frontmatter (7 agents, 7 prompts)
* 70 Pester tests cover both validation and refresh logic
* Catalog currently tracks 24 models across 5 tiers (free, fast, standard, premium, ultra)


### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by katriendg**
## Description

This PR introduces **per-agent model selection** via frontmatter, backed by a validated model catalog that tracks GitHub Copilot's evolving model lineup. Simple tasks (git operations, issue triage, research) now route to fast-tier models at a fraction of the cost, while complex agents inherit the session model for full capability.

Additionally, this PR removes the self-referential `/compact` handoff from 12 agents, eliminating the root cause of **Autopilot infinite loops** reported in #1420. The disk-first `.copilot-tracking/` architecture and Memory Agent already provide equivalent persistence without the loop risk.

### Model Selection Infrastructure

> Cost-first principle: use fast models for read-only research and validation; inherit session model for code generation and complex reasoning.

- Added **model catalog** (`scripts/linting/model-catalog.json`) tracking 25 models across 5 tiers (free, fast, standard, premium, ultra) with multiplier values, vendor attribution, and GA/preview/retiring status
- Added **catalog refresh script** (`scripts/linting/Update-ModelCatalog.ps1`) that fetches authoritative YAML from `github/docs` for model release status and multiplier data; marks removed models as *retiring* with 60-day grace period rather than deleting
- Added **validation script** (`scripts/linting/Test-ModelReferences.ps1`) that scans all `.agent.md` and `.prompt.md` files for model frontmatter and validates references against the catalog; reports invalid models as errors, retiring models as warnings
- Added **JSON schema** (`scripts/linting/schemas/model-catalog.schema.json`) for structural validation of the catalog file
- Added **weekly CI workflow** (`.github/workflows/model-validation.yml`) running every Wednesday plus PR-triggered validation on agent/prompt/catalog changes; includes catalog freshness check and artifact upload
- Integrated `lint:models` and `lint:models:refresh` into `package.json`; model validation runs as part of the `lint:all` chain

### Per-Agent and Per-Prompt Model Assignment

Assigned fast-tier models to 7 **subagents** performing read-heavy validation tasks: *researcher-subagent*, *plan-validator*, *implementation-validator*, *prompt-evaluator*, *rpi-validator*, *codebase-profiler*, and *report-generator*. Each declares a prioritized fallback array: `Claude Haiku 4.5 → GPT-5.4 mini`.

Assigned `Claude Haiku 4.5 (copilot)` to 7 **prompts** handling mechanical operations: *git-commit-message*, *git-commit*, *git-setup*, *github-add-issue*, *github-discover-issues*, *github-triage-issues*, and *checkpoint*.

Added "Model Selection for Subagents" guidance to 6 **parent agents** (*task-researcher*, *task-planner*, *task-implementor*, *task-reviewer*, *prompt-builder*, *security-reviewer*) documenting cost-first dispatch decisions and VS Code tier constraint behavior.

### /compact Handoff Removal (Fixes #1420)

Removed the `Compact` handoff entry from all 12 agents where it appeared. Eleven had it as their **first handoff**, causing Autopilot to auto-execute it on every turn completion, creating an infinite self-referential loop.

- Updated `rai-identity.instructions.md` to remove the "Compact handoff" exit point reference from disclaimer display logic
- Updated `docs/rpi/context-engineering.md` to recommend `/checkpoint` (Memory Agent) for cross-phase persistence and clarify that `/compact` remains available as a typed command

PR #1492 (`feat/context-working`) adds Context Discipline to 5 RPI parent agents, enforcing disk-first lean responses. The `/compact` handoff is now architecturally redundant because:

1. **Disk-first `.copilot-tracking/` files** — all state already lives on disk
2. **Memory Agent** — provides structured session persistence with handoff to a different agent (non-looping)
3. **PR #1492 Context Discipline** — caps subagent responses to executive summaries, reducing context bloat at the source

### Test Coverage

- Added 41 Pester tests (`Test-ModelReferences.Tests.ps1`) covering validation logic, frontmatter parsing, and error handling
- Added 29 Pester tests (`Test-UpdateModelCatalog.Tests.ps1`) covering catalog merge, comparison, and refresh logic

## Related Issue(s)

Fixes #1420
Closes #1540

## Type of Change

Select all that apply:

**Code & Documentation:**

* [ ] Bug fix (non-breaking change fixing an issue)
* [x] New feature (non-breaking change adding functionality)
* [x] Breaking change (fix or feature causing existing functionality to change)
* [x] Documentation update

**Infrastructure & Configuration:**

* [x] GitHub Actions workflow
* [ ] Linting configuration (markdown, PowerShell, etc.)
* [ ] Security configuration
* [ ] DevContainer configuration
* [ ] Dependency update

**AI Artifacts:**

* [x] Reviewed contribution with `prompt-builder` agent and addressed all feedback
* [x] Copilot instructions (`.github/instructions/*.instructions.md`)
* [x] Copilot prompt (`.github/prompts/*.prompt.md`)
* [x] Copilot agent (`.github/agents/*.agent.md`)
* [ ] Copilot skill (`.github/skills/*/SKILL.md`)

> Note for AI Artifact Contributors:
>
> * Agents: Research, indexing/referencing other project (using standard VS Code GitHub Copilot/MCP tools), planning, and general implementation agents likely already exist. Review `.github/agents/` before creating new ones.
> * Skills: Must include both bash and PowerShell scripts. See [Skills](../docs/contributing/skills.md).
> * Model Versions: Only contributions targeting the **latest Anthropic and OpenAI models** will be accepted. Older model versions (e.g., GPT-3.5, Claude 3) will be rejected.
> * See [Agents Not Accepted](../docs/contributing/custom-agents.md#agents-not-accepted) and [Model Version Requirements](../docs/contributing/ai-artifacts-common.md#model-version-requirements).

**Other:**

* [x] Script/automation (`.ps1`, `.sh`, `.py`)
* [ ] Other (please describe):

## Sample Prompts (for AI Artifact Contributions)

**User Request:**

Invoke any RPI agent (e.g., `task researcher`) with a research task. The agent dispatches its Researcher Subagent at fast-tier cost automatically. Run `npm run lint:models` to validate all model references.

**Execution Flow:**

1. Parent agent evaluates task type (read-only vs code-generation)
2. For research/validation tasks, parent specifies `model: "Claude Haiku 4.5 (copilot)"` on `runSubagent` call
3. VS Code resolves model against cost tier constraint (cannot exceed parent model tier)
4. Subagent executes at fast-tier cost; results written to `.copilot-tracking/` disk files
5. If tier constraint blocks downgrade, platform falls back to session model gracefully

**Output Artifacts:**

- `logs/model-validation-results.json` — structured validation results with per-file status
- `scripts/linting/model-catalog.json` — refreshed catalog after `lint:models:refresh`

**Success Indicators:**

- `npm run lint:models` exits 0 with no invalid model references
- Subagent invocations show model name in VS Code chat header when explicitly set
- No Autopilot infinite loops when agents complete their work

## Testing

- `npm run lint:models` — model reference validation (validates all 14 model-annotated files)
- Security analysis: no sensitive data exposure, no privilege escalation, workflow uses read-only permissions
- Diff-based assessment: all changes are configuration-level (frontmatter, handoff entries, guidance sections); no business logic modified
- Manual testing performed

> [!NOTE]
> Add manual testing descriptions when applicable.

### Required Checks

* [x] Documentation is updated (if applicable)
* [x] Files follow existing naming conventions
* [ ] Changes are backwards compatible (if applicable)
* [x] Tests added for new functionality (if applicable)

### AI Artifact Contributions

* [x] Used `/prompt-analyze` to review contribution
* [x] Addressed all feedback from `prompt-builder` review
* [x] Verified contribution follows common standards and type-specific requirements

### Required Automated Checks

The following validation commands must pass before merging:

* [x] Markdown linting: `npm run lint:md`
* [x] Spell checking: `npm run spell-check`
* [x] Frontmatter validation: `npm run lint:frontmatter`
* [x] Skill structure validation: `npm run validate:skills`
* [x] Link validation: `npm run lint:md-links`
* [x] PowerShell analysis: `npm run lint:ps`
* [x] Plugin freshness: `npm run plugin:generate`
* [x] Docusaurus tests: `npm run docs:test`

## Security Considerations

* [x] This PR does not contain any sensitive or NDA information
* [ ] Any new dependencies have been reviewed for security issues (N/A — no new runtime dependencies added)
* [x] Security-related scripts follow the principle of least privilege

> [!WARNING]
> This PR includes **experimental** GHCP artifacts that may have breaking changes.
> - `.github/agents/hve-core/task-challenger.agent.md`
> - `.github/agents/experimental/experiment-designer.agent.md`
> - `.github/agents/experimental/pptx.agent.md`
> - `.github/agents/security/security-planner.agent.md`
> - `.github/agents/security/sssc-planner.agent.md`
> - `.github/agents/security/security-reviewer.agent.md`
> - `.github/agents/security/subagents/codebase-profiler.agent.md`
> - `.github/agents/security/subagents/report-generator.agent.md`
> - `.github/agents/rai-planning/rai-planner.agent.md`

## GHCP Artifact Maturity

| File | Type | Maturity | Notes |
|------|------|----------|-------|
| `.github/agents/hve-core/rpi-agent.agent.md` | Agent | ✅ stable | All builds |
| `.github/agents/hve-core/task-researcher.agent.md` | Agent | ✅ stable | All builds |
| `.github/agents/hve-core/task-planner.agent.md` | Agent | ✅ stable | All builds |
| `.github/agents/hve-core/task-implementor.agent.md` | Agent | ✅ stable | All builds |
| `.github/agents/hve-core/task-reviewer.agent.md` | Agent | ✅ stable | All builds |
| `.github/agents/hve-core/prompt-builder.agent.md` | Agent | ✅ stable | All builds |
| `.github/agents/hve-core/task-challenger.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/experimental/experiment-designer.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/experimental/pptx.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/security/security-planner.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/security/sssc-planner.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/security/security-reviewer.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/security/subagents/codebase-profiler.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/security/subagents/report-generator.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/agents/rai-planning/rai-planner.agent.md` | Agent | ⚠️ experimental | Pre-release only |
| `.github/prompts/hve-core/checkpoint.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/prompts/hve-core/git-commit-message.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/prompts/hve-core/git-commit.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/prompts/hve-core/git-setup.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/prompts/github/github-add-issue.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/prompts/github/github-discover-issues.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/prompts/github/github-triage-issues.prompt.md` | Prompt | ✅ stable | All builds |
| `.github/instructions/rai-planning/rai-identity.instructions.md` | Instructions | ⚠️ experimental | Pre-release only |

### GHCP Maturity Acknowledgment
- [ ] I acknowledge this PR includes non-stable GHCP artifacts
- [ ] Non-stable artifacts are intentional for this change

## Additional Notes

- The `/compact` removal is a **breaking change** for users who relied on the handoff button. The `/compact` typed command remains available; only the agent-surfaced handoff is removed.
- Model catalog currently tracks 25 models; the automated refresh runs weekly to catch additions, removals, and multiplier changes from GitHub's upstream YAML sources.
- The VS Code cost tier constraint means subagents can only use models at the same or lower tier than the parent. All guidance sections document this limitation and the graceful fallback behavior.

## Follow-up Tasks

- Monitor weekly CI workflow for first catalog drift detection to confirm automation works end-to-end
- Consider extending model selection to remaining prompts (pull-request, doc-ops) once cost savings are validated

### CODE REVIEW

File: .github/workflows/model-validation.yml
Comment (bindsi) [inline .github/workflows/model-validation.yml:1]: If I understand right, manual intervention is required in case there is a model mismatch, correct?

**Comment by bindsi** [comment]
## Review

Two bundled changes: per-agent/prompt `model` frontmatter backed by a validated catalog with weekly refresh, and removal of the self-referential `/compact` handoff from 12 agents to fix Autopilot loops (#1420). All 59 CI checks green.

### Strengths

- **Well-scoped and well-justified.** PR body cleanly separates the two threads. The architectural argument for `/compact` removal (disk-first `.copilot-tracking/` + Memory Agent + #1492 Context Discipline making it redundant) is sound.
- **Catalog-driven validation.** JSON schema, refresh script that fetches authoritative YAML from `github/docs`, retiring grace period instead of hard delete — right pattern.
- **Strong test coverage.** 70 Pester tests covering frontmatter parsing, validation logic, tier classification, and catalog comparison. Edge cases (Not applicable multiplier, missing multiplier entry, mixed invalid/retiring) are exercised.
- **Documentation is thorough and coherent** across `ai-artifacts-common.md`, `custom-agents.md`, `prompts.md`, and `context-engineering.md`.
- **Cost-first guidance is consistent** across parent agents and follows a clear pattern: read-only/validation → fast tier; code generation/architecture → inherit session model.

### Concerns

**1. Silent catalog drift in CI (medium).** `model-validation.yml` runs `Update-ModelCatalog.ps1` *before* validation, but the refreshed catalog isn't committed. PRs validate against a *fresher* catalog than what's in the repo — a reference can pass CI because the model exists upstream while the committed `model-catalog.json` is stale. The "Report catalog drift" step only flags retiring entries; it doesn't surface added/removed/multiplier-changed models that appear only in the just-fetched copy.

Suggestion: either (a) validate against the committed catalog only and run refresh on a separate scheduled job that opens an auto-PR, or (b) fail the CI step when the in-memory refreshed catalog differs materially from the committed one.

**2. Provider-allowlist not mechanically enforced (medium).** The catalog correctly contains non-Anthropic/OpenAI entries (`Goldeneye`, `Raptor mini`, `Gemini 3 Flash`, `Grok Code Fast 1`) pulled from upstream YAML. `lint:models` accepts them, but they'd violate the "Anthropic and OpenAI only" policy in `ai-artifacts-common.md`. Consider an additional provider-allowlist check so policy is enforced, not just documented.

**3. Upstream pin (low).** `Update-ModelCatalog.ps1` fetches from `github/docs@main`. Combined with #1, PR validation results are non-deterministic across runs. Worth at least a script comment.

**4. Workflow runner pin (low — verify).** `model-validation.yml` uses `runs-on: ubuntu-latest`. Repo conventions typically require a pinned runner version (e.g., `ubuntu-24.04`).

**5. Untested branch in `Update-ModelCatalog.ps1` (low).** The "mark removed as retiring" overlay logic in the main block (~lines 245–265) — 60-day grace transition and update-multiplier-on-existing-entry path — has no tests. `Compare-Catalogs` itself is well covered.

**6. `Get-FrontmatterFromFile` regex (low).** `'(?s)^---\r?\n(.+?)\r?\n---'` doesn't anchor the closing `---` to a line boundary. Lazy `.+?` keeps it correct for current files, but worth aligning with the existing frontmatter validator's pattern.

**7. Breaking change disclosure.** `/compact` handoff removal is correctly called out as breaking in the PR body. Verify release-please/changelog surfaces this (e.g., `feat!:` or `BREAKING CHANGE:` footer) so downstream consumers see it.

**8. Minor.** `Test-ModelReferences.ps1` uses `(Get-Location).Path` for relative-path computation — awkward when invoked outside repo root. Consider basing on `$PSScriptRoot`.

### Verdict

Approve with non-blocking suggestions. Core mechanics and tests are solid. **#1 (silent catalog drift)** and **#2 (provider-allowlist)** are the most actionable — both undermine the validation guarantee the PR is selling. Worth addressing before merge or in immediate follow-up.

**Review by bindsi — APPROVED** [review]
Really like it, that´s such an improvement and optimisation of cost impact for users. Thank you so much

File: scripts/linting/model-catalog.json [hunk 1]
```diff
@@ -0,0 +1,152 @@
+{
```
[the remark this answers is not in the document]
Comment (katriendg) [reply scripts/linting/model-catalog.json:1]: this is not true, schema is there just not as the second line

File: scripts/linting/Update-ModelCatalog.ps1 [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+                # Update multiplier if changed
+                $change = $diff.changed | Where-Object { $_.name -eq $curr.name }
+                if ($change) {
+                    $curr.multiplier = $change.newMultiplier
+                }
+                $finalModels += $curr
+            }
+        }
+        # Add new models
+        $finalModels += $diff.added
```
[the remark this answers is not in the document]
Comment (katriendg) [reply scripts/linting/Update-ModelCatalog.ps1:332]: fixing

File: .github/workflows/model-validation.yml
Comment (katriendg) [reply .github/workflows/model-validation.yml:1]: It's going to be an exercise of testing and looking at impact. I've been testing it manually and validating the Haiku gets called appropriately and that is a great cost reduction. But there is no automated comparison. My 'testing' is the results of the agent work/prompt work, and using the Copilot chat debug logs and views ;)

**Comment by katriendg** [comment]
> ## Review
> Two bundled changes: per-agent/prompt `model` frontmatter backed by a validated catalog with weekly refresh, and removal of the self-referential `/compact` handoff from 12 agents to fix Autopilot loops (#1420). All 59 CI checks green.
> ... omitting the rest

Thanks @bindsi for your review. I believe we have addressed everything:

# PR Review Comments — Resolution Summary

| # | Status | Action |
|---|--------|--------|
| 1 | Implemented | Added "Detect catalog drift" CI step in `model-validation.yml` |
| 2 | Implemented | Added `provider` field, `providerAllowlist`, and enforcement |
| 3 | Implemented | Added `.NOTES` comment documenting upstream non-determinism |
| 4 | No action | `ubuntu-latest` is the required pattern per repo conventions |
| 5 | No action | Tests already exist for the retiring/multiplier-update paths |
| 6 | Implemented | Aligned the frontmatter regex patterns to `(.*?)\r?\n---(\r?\n|\z)` |
| 7 | No action | Commit-message - ensure `feat!:` on merge (updated) |
| 8 | Implemented | Replaced `(Get-Location).Path` with `$RepoRoot` via `git rev-parse` |

File: .github/workflows/model-validation.yml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+jobs:
+  validate-models:
+    name: Validate Model References
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
```
[the remark this answers is not in the document]
Comment (katriendg) [reply .github/workflows/model-validation.yml:30]: SHA de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2 is used consistently across 20+ workflow files in this repo (markdown-lint, pester-tests, scorecard, copilot-setup-steps, etc.). This is the correct repo-wide pinned version. The comment is a false positive.

File: .github/workflows/model-validation.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      - name: Validate model references against catalog
+        id: validate
+        shell: pwsh
+        run: |
+          New-Item -ItemType Directory -Force -Path logs | Out-Null
+          & scripts/linting/Test-ModelReferences.ps1 -OutputPath logs/model-validation-results.json
+
+      - name: Upload results
+        if: always()
+        uses: actions/upload-artifact@043fb46d1a93c77aae656e7c1c64a875d1fc6a0a # v7.0.1
```
[the remark this answers is not in the document]
Comment (katriendg) [reply .github/workflows/model-validation.yml:64]: false

File: .github/workflows/model-validation.yml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    paths:
+      - '.github/agents/**/*.agent.md'
+      - '.github/prompts/**/*.prompt.md'
+      - 'scripts/linting/model-catalog.json'
+      - 'scripts/linting/Test-ModelReferences.ps1'
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (katriendg) [reply .github/workflows/model-validation.yml:17]: skipping, unlikely

File: scripts/linting/Update-ModelCatalog.ps1 [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+    .PARAMETER Url
+    URL to fetch.
+
+    .OUTPUTS
+    Parsed YAML content.
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [string]$Url
```
[the remark this answers is not in the document]
Comment (katriendg) [reply scripts/linting/Update-ModelCatalog.ps1:75]: fixing

File: .github/agents/hve-core/subagents/researcher-subagent.agent.md [hunk 7]
```diff
@@ -2,6 +2,9 @@
 name: Researcher Subagent
 description: 'Research subagent using search tools, read tools, fetch web page, github repo, and mcp tools'
 user-invocable: false
+model:
+  - Claude Haiku 4.5 (copilot)
+  - GPT-5.4 mini (copilot)
```
[the remark this answers is not in the document]
Comment (katriendg) [reply .github/agents/hve-core/subagents/researcher-subagent.agent.md:7]: updated

**Review by WilliamBerryiii — APPROVED** [review]
